### ISSUE TITLE
损失变化过程提问

### ISSUE THREAD

**Issue by Edwards1997** [Shawn-Yu-1/AGGNet#1]
您好，我想问一下我一直不理解这个代码中判别器损失和生成器损失正常的变化过程，希望您可以赐教下

**Comment by Shawn-Yu-1** [comment]
你是想问GAN训练的问题吗？一般生成器损失是一直降的，鉴别器损失可以不用管。总之不能让生成器损失来回跳。而且我的鉴别器很简单，你可以试着换个更好的，有人用我的作为baseline,换了个鉴别器，发了一篇了。

**Comment by Edwards1997** [comment]
> 你是想问GAN训练的问题吗？一般生成器损失是一直降的，鉴别器损失可以不用管。总之不能让生成器损失来回跳。而且我的鉴别器很简单，你可以试着换个更好的，有人用我的作为baseline,换了个鉴别器，发了一篇了。
对的我想问的就是GAN训练的问题，我跑特定类别的图像修复任务的时候生成器对抗损失是一直在正负横跳的，而生成器总损失是一直在下降接近0，判别器的损失就一直保持在2左右，对于我的任务效果是出奇的好，我就是不太懂这个GAN网络训练的损失变化过程，谢谢大佬赐教。

**Comment by Edwards1997** [comment]
> 你是想问GAN训练的问题吗？一般生成器损失是一直降的，鉴别器损失可以不用管。总之不能让生成器损失来回跳。而且我的鉴别器很简单，你可以试着换个更好的，有人用我的作为baseline,换了个鉴别器，发了一篇了。

对的，我就是想询问GAN网络训练的过程，我在执行特定图像的修复任务中这个作为baseline展现出出色的效果，但我不是很理解训练过程中的这些变化，我的生成器总损失是在下降的，而生成器的对抗损失出现了正负横跳的现象，请大佬赐教

**Comment by Shawn-Yu-1** [comment]
GAN网络想训练好，一定不能让生成器的对抗损失来回跳，你可以试着降低鉴别器的学习率，让鉴别器学的差一点，我这个工作中鉴别器学习率是生成器的1/10.

**Comment by Shawn-Yu-1** [comment]
当然还可以加一点其他loss去辅助学习，会有很明显的提升。

**Comment by Edwards1997** [comment]
好的，谢谢大佬

**Comment by Edwards1997** [comment]
大佬，我又来提问了，我将全部的注意力机制去除并将其改为原本的门控卷积跟原有的注意力模型对比，指标基本上没有什么变化，这个是正常的吗？而且我试过很多模块，并且将大核注意力改成了多尺度的发现效果只是微乎其微，现在又陷入了停滞，想请大佬赐教下

**Comment by Edwards1997** [comment]
还有一个奇怪的现象，使用去除所有注意力的模型的前期修复效果已经比别人后期的修复效果更佳

**Comment by Shawn-Yu-1** [comment]
> 还有一个奇怪的现象，使用去除所有注意力的模型的前期修复效果已经比别人后期的修复效果更佳

你可以描述以下你做的任务吗？你这前期后期我也不知道是啥。

**Comment by Shawn-Yu-1** [comment]
> 大佬，我又来提问了，我将全部的注意力机制去除并将其改为原本的门控卷积跟原有的注意力模型对比，指标基本上没有什么变化，这个是正常的吗？而且我试过很多模块，并且将大核注意力改成了多尺度的发现效果只是微乎其微，现在又陷入了停滞，想请大佬赐教下

去除注意力的话，我可以理解为deepfillv2吗？我这个要比他好很多。如果你不用轮廓图的话，注意力机制应该没啥用。或者你训练会不会又啥问题。

**Comment by Edwards1997** [comment]
我做的是虹膜的图像遮掩修复，我自己引入了一个纹理图作为先验
class GatedAttentionNetV1(nn.Module):
    def __init__(self, in_channels=8, out_channels=3, inner_channels=[64, 128, 256, 512], num_heads=[2, 2, 2, 2],
                 atten_type="sparse", last_covn=32, embed=4):
        super().__init__()
        self.inconv = GatedConv(8, 32, 7, 1, 3)
        self.encoder0 = GatedConv(32, 64, 5, 2, 2)
        self.encoder1 = GatedConv(64, 128, 5, 2, 2)
        self.encoder2 = GatedConv(128, 256, 4, 2, 1)
        self.encoder3 = GatedConv(256, 512, 4, 2, 1)
        self.embedding = nn.Sequential(
            GatedConv(512, 512, 3, 1, 1),
            GatedConv(512, 512, 3, 1, 1),
            GatedConv(512, 512, 3, 1, 1),
            GatedConv(512, 512, 3, 1, 1),
            GatedConv(512, 512, 3, 1, 1)
        )

        self.decoder0 = GatedConv(1024, 256, 3, 1, 1)
        self.upsample0 = nn.Upsample(scale_factor=2, mode='nearest')
        self.decoder1 = GatedConv(512, 128, 3, 1, 1)
        self.upsample1 = nn.Upsample(scale_factor=2, mode='nearest')
        self.decoder2 = GatedConv(256, 64, 5, 1, 2)
        self.upsample2 = nn.Upsample(scale_factor=2, mode='nearest')
        self.decoder3 = GatedConv(128, 32, 5, 1, 2)
        self.upsample3 = nn.Upsample(scale_factor=2, mode='nearest')

        self.outcovn = nn.Conv2d(32, 3, 7, 1, 3)

    def forward(self, img, edge, texture, mask):
        x = torch.cat([img * (1 - mask), edge * (1 - mask), texture * (1 - mask), mask - 0.5], dim=1)

        out1 = self.encoder0(self.inconv(x))
        out2 = self.encoder1(out1)
        out3 = self.encoder2(out2)
        out4 = self.encoder3(out3)
        out = self.embedding(out4)
        # 512+512 -> 256
        out = self.decoder0(torch.cat([out, out4], dim=1))
        out = self.upsample0(out)
        # 256+256 -> 128
        out = self.decoder1(torch.cat([out, out3], dim=1))
        out = self.upsample1(out)
        # 128+128 -> 64
        out = self.decoder2(torch.cat([out, out2], dim=1))
        out = self.upsample2(out)
        # 64+64 -> 32
        out = self.decoder3(torch.cat([out, out1], dim=1))
        out = self.upsample3(out)
        out = self.outcovn(out)
        out = out * mask + img * (1 - mask)
        return out

**Comment by Shawn-Yu-1** [comment]
> 我做的是虹膜的图像遮掩修复，我自己引入了一个纹理图作为先验
> class GatedAttentionNetV1(nn.Module):
>     def __init__(self, in_channels=8, out_channels=3, inner_channels=[64, 128, 256, 512], num_heads=[2, 2, 2, 2],
>                  atten_type="sparse", last_covn=32, embed=4):
>         super().__init__()
>         self.inconv = GatedConv(8, 32, 7, 1, 3)
>         self.encoder0 = GatedConv(32, 64, 5, 2, 2)
>         self.encoder1 = GatedConv(64, 128, 5, 2, 2)
>         self.encoder2 = GatedConv(128, 256, 4, 2, 1)
>         self.encoder3 = GatedConv(256, 512, 4, 2, 1)
>         self.embedding = nn.Sequential(
>             GatedConv(512, 512, 3, 1, 1),
>             GatedConv(512, 512, 3, 1, 1),
>             GatedConv(512, 512, 3, 1, 1),
>             GatedConv(512, 512, 3, 1, 1),
>             GatedConv(512, 512, 3, 1, 1)
>         )
> 
>         self.decoder0 = GatedConv(1024, 256, 3, 1, 1)
>         self.upsample0 = nn.Upsample(scale_factor=2, mode='nearest')
>         self.decoder1 = GatedConv(512, 128, 3, 1, 1)
>         self.upsample1 = nn.Upsample(scale_factor=2, mode='nearest')
>         self.decoder2 = GatedConv(256, 64, 5, 1, 2)
>         self.upsample2 = nn.Upsample(scale_factor=2, mode='nearest')
>         self.decoder3 = GatedConv(128, 32, 5, 1, 2)
>         self.upsample3 = nn.Upsample(scale_factor=2, mode='nearest')
> 
>         self.outcovn = nn.Conv2d(32, 3, 7, 1, 3)
> 
>     def forward(self, img, edge, texture, mask):
>         x = torch.cat([img * (1 - mask), edge * (1 - mask), texture * (1 - mask), mask - 0.5], dim=1)
> 
>         out1 = self.encoder0(self.inconv(x))
>         out2 = self.encoder1(out1)
>         out3 = self.encoder2(out2)
>         out4 = self.encoder3(out3)
>         out = self.embedding(out4)
>         # 512+512 -> 256
>         out = self.decoder0(torch.cat([out, out4], dim=1))
>         out = self.upsample0(out)
>         # 256+256 -> 128
>         out = self.decoder1(torch.cat([out, out3], dim=1))
>         out = self.upsample1(out)
>         # 128+128 -> 64
>         out = self.decoder2(torch.cat([out, out2], dim=1))
>         out = self.upsample2(out)
>         # 64+64 -> 32
>         out = self.decoder3(torch.cat([out, out1], dim=1))
>         out = self.upsample3(out)
>         out = self.outcovn(out)
>         out = out * mask + img * (1 - mask)
>         return out

虹膜图像可能轮廓不够丰富，是有可能出现这种情况。你可以看一下你抽取的先验图是不是没啥东西，想办法增强先验的内容。至于本模型前期就效果好，这个主要是本模型收敛比较快。 还有用于先验的图，我建议用灰度图就行了，不需要rgb。 而且如果数据集纹理比较单一的话，通道也不需要这么高，256可能就够用了。

**Comment by Shawn-Yu-1** [comment]
对于医学图像的话，数据增强是很有用的，你可以考虑加入一些新的数据增强方法。

**Comment by Edwards1997** [comment]
> 对于医学图像的话，数据增强是很有用的，你可以考虑加入一些新的数据增强方法。

好的，我这个就是单纯的虹膜图像修复，谢谢大佬，我有点思路了

**Comment by Edwards1997** [comment]
我现在一个数据集训练的epoch是300轮，训练集为1800张，验证集是450，img_szie 为256 但是我突然发现用这个模型，在300轮的时候达到的效果跟加了注意力机制的一样，这种情况是没有收敛吗

> > 我做的是虹膜的图像遮掩修复，我自己引入了一个纹理图作为先验 class GatedAttentionNetV1（nn.模块）： def **init**（self， in_channels=8， out_channels=3， inner_channels=[64， 128， 256， 512]， num_heads=[2， 2， 2]， atten_type=“sparse”， last_covn=32， embed=4）： super（）.**init**（） self.inconv = GatedConv（8， 32， 7， 1， 3） self.encoder0 = GatedConv（32， 64， 5， 2， 2） self.encoder1 = GatedConv（64， 128， 5， 2， 2） self.encoder2 = GatedConv（128， 256， 4， 2， 1） self.encoder3 = GatedConv（256， 512， 4， 2， 1） self.embedding = nn.顺序（ GatedConv（512， 512， 3， 1， 1）， GatedConv（512， 512， 3， 1， 1）， GatedConv（512， 512， 3， 1， 1）， GatedConv（512， 512， 3， 1， 1）， GatedConv（512， 512， 3， 1， 1）
> > )
> > ```
> >     self.decoder0 = GatedConv(1024, 256, 3, 1, 1)
> >     self.upsample0 = nn.Upsample(scale_factor=2, mode='nearest')
> >     self.decoder1 = GatedConv(512, 128, 3, 1, 1)
> >     self.upsample1 = nn.Upsample(scale_factor=2, mode='nearest')
> >     self.decoder2 = GatedConv(256, 64, 5, 1, 2)
> >     self.upsample2 = nn.Upsample(scale_factor=2, mode='nearest')
> >     self.decoder3 = GatedConv(128, 32, 5, 1, 2)
> >     self.upsample3 = nn.Upsample(scale_factor=2, mode='nearest')
> > 
> >     self.outcovn = nn.Conv2d(32, 3, 7, 1, 3)
> > 
> > def forward(self, img, edge, texture, mask):
> >     x = torch.cat([img * (1 - mask), edge * (1 - mask), texture * (1 - mask), mask - 0.5], dim=1)
> > 
> >     out1 = self.encoder0(self.inconv(x))
> >     out2 = self.encoder1(out1)
> >     out3 = self.encoder2(out2)
> >     out4 = self.encoder3(out3)
> >     out = self.embedding(out4)
> >     # 512+512 -> 256
> >     out = self.decoder0(torch.cat([out, out4], dim=1))
> >     out = self.upsample0(out)
> >     # 256+256 -> 128
> >     out = self.decoder1(torch.cat([out, out3], dim=1))
> >     out = self.upsample1(out)
> >     # 128+128 -> 64
> >     out = self.decoder2(torch.cat([out, out2], dim=1))
> >     out = self.upsample2(out)
> >     # 64+64 -> 32
> >     out = self.decoder3(torch.cat([out, out1], dim=1))
> >     out = self.upsample3(out)
> >     out = self.outcovn(out)
> >     out = out * mask + img * (1 - mask)
> >     return out
> > ```
> 
> 虹膜图像可能轮廓不够丰富，是有可能出现这种情况。你可以看一下你抽取的先验图是不是没啥东西，想办法增强先验的内容。至于本模型前期就效果好，这个主要是本模型收敛比较快。还有用于先验的图，我建议用灰度图就行了，不需要rgb。而且如果数据集纹理比较单一的话，通道也不需要这么高，256可能就够用了。

**Comment by Shawn-Yu-1** [comment]
> 我现在一个数据集训练的epoch是300轮，训练集为1800张，验证集是450，img_szie 为256 但是我突然发现用这个模型，在300轮的时候达到的效果跟加了注意力机制的一样，这种情况是没有收敛吗
> 
> > > 我做的是虹膜的图像遮掩修复，我自己引入了一个纹理图作为先验 class GatedAttentionNetV1（nn.模块）： def **init**（self， in_channels=8， out_channels=3， inner_channels=[64， 128， 256， 512]， num_heads=[2， 2， 2]， atten_type=“sparse”， last_covn=32， embed=4）： super（）.**init**（） self.inconv = GatedConv（8， 32， 7， 1， 3） self.encoder0 = GatedConv（32， 64， 5， 2， 2） self.encoder1 = GatedConv（64， 128， 5， 2， 2） self.encoder2 = GatedConv（128， 256， 4， 2， 1） self.encoder3 = GatedConv（256， 512， 4， 2， 1） self.embedding = nn.顺序（ GatedConv（512， 512， 3， 1， 1）， GatedConv（512， 512， 3， 1， 1）， GatedConv（512， 512， 3， 1， 1）， GatedConv（512， 512， 3， 1， 1）， GatedConv（512， 512， 3， 1， 1）
> > > )
> > > ```
> > >     self.decoder0 = GatedConv(1024, 256, 3, 1, 1)
> > >     self.upsample0 = nn.Upsample(scale_factor=2, mode='nearest')
> > >     self.decoder1 = GatedConv(512, 128, 3, 1, 1)
> > >     self.upsample1 = nn.Upsample(scale_factor=2, mode='nearest')
> > >     self.decoder2 = GatedConv(256, 64, 5, 1, 2)
> > >     self.upsample2 = nn.Upsample(scale_factor=2, mode='nearest')
> > >     self.decoder3 = GatedConv(128, 32, 5, 1, 2)
> > >     self.upsample3 = nn.Upsample(scale_factor=2, mode='nearest')
> > > 
> > >     self.outcovn = nn.Conv2d(32, 3, 7, 1, 3)
> > > 
> > > def forward(self, img, edge, texture, mask):
> > >     x = torch.cat([img * (1 - mask), edge * (1 - mask), texture * (1 - mask), mask - 0.5], dim=1)
> > > 
> > >     out1 = self.encoder0(self.inconv(x))
> > >     out2 = self.encoder1(out1)
> > >     out3 = self.encoder2(out2)
> > >     out4 = self.encoder3(out3)
> > >     out = self.embedding(out4)
> > >     # 512+512 -> 256
> > >     out = self.decoder0(torch.cat([out, out4], dim=1))
> > >     out = self.upsample0(out)
> > >     # 256+256 -> 128
> > >     out = self.decoder1(torch.cat([out, out3], dim=1))
> > >     out = self.upsample1(out)
> > >     # 128+128 -> 64
> > >     out = self.decoder2(torch.cat([out, out2], dim=1))
> > >     out = self.upsample2(out)
> > >     # 64+64 -> 32
> > >     out = self.decoder3(torch.cat([out, out1], dim=1))
> > >     out = self.upsample3(out)
> > >     out = self.outcovn(out)
> > >     out = out * mask + img * (1 - mask)
> > >     return out
> > > ```
> > 
> > 虹膜图像可能轮廓不够丰富，是有可能出现这种情况。你可以看一下你抽取的先验图是不是没啥东西，想办法增强先验的内容。至于本模型前期就效果好，这个主要是本模型收敛比较快。还有用于先验的图，我建议用灰度图就行了，不需要rgb。而且如果数据集纹理比较单一的话，通道也不需要这么高，256可能就够用了。
> 
> 
收敛应该是收敛了，但是感觉数据集太小了，没必要训这门多epoch。 效果差不多只能说你的任务简单用不到先验，或者先验信息太少了。

**Comment by Edwards1997** [comment]
好的，我再摸索下，谢谢大佬

**Comment by Edwards1997** [comment]
大佬，我又来了，
我总感觉我的任务比较简单，我把结构灰度图去除得到的结果对比基本也是一致的，我想使用这个原模型进行一个轻量化的处理，大佬觉得有没有可行性。
